src/apple/alohalytics_objc.mm

```diff
@@ -401,12 +401,27 @@ + (void)setup:(NSArray *)serverUrls withLaunchOptions:(NSDictionary *)options {
   instance.LogEvent("$launch", params);
 #if (TARGET_OS_IPHONE > 0)
   // Initialize User-Agent asynchronously and log additional system info for iOS, as it takes significant time at startup.
-  dispatch_async(dispatch_get_main_queue(), ^{
-    gBrowserUserAgent = [[[UIWebView alloc] initWithFrame:CGRectZero] stringByEvaluatingJavaScriptFromString:@"navigator.userAgent"];
-    if (shouldSendUpdatedSystemInformation) {
-      LogSystemInformation(gBrowserUserAgent);
-    }
-  });
+    dispatch_async(dispatch_get_main_queue(), ^{
+        WKWebView *webView = [[WKWebView alloc] initWithFrame:CGRectZero];
+        [webView evaluateJavaScript: @"navigator.userAgent"
+                  completionHandler:^(id _Nullable result, NSError * _Nullable error) {
+            if (result != nil && [result isKindOfClass:[NSString class]]) {
+                gBrowserUserAgent = (NSString *)result;
+                if (shouldSendUpdatedSystemInformation) {
+                    LogSystemInformation(gBrowserUserAgent);
+                }
+            } else if (error != nil) {
+                NSLog(@"Alohalytics ERROR: can not get userAgent: %@", error);
+            }
+            dispatch_async(dispatch_get_main_queue(), ^{
+                // we need reference to webView to keep it alive until
+                // this moment to prevent errors and assertions
+                // if webView would be destroyed before `completionHandler`
+                // would be called or right after `completionHandler` finished
+                WKWebView *keepAlive __attribute__((unused)) = webView;
```

Comment:
Can this hack be replaced by something like `webView = nil` in the completion handler?

Reply:
Yes and no. It should not be in `completionHandler`. It should be queued into some future moment.
Because of if
```
completionHandler:^(id _Nullable result, NSError * _Nullable error) {
   //main code
   somehowClearWebViewHere(); 
}
```
then you got bunch of messages like "assert failed" and so on.
Looks like `webView` do something after calling `completionHandler`,
and it don't like if `self` is `nil`.

It can be done like this:
```
dispatch_async(^{
__block  WKWebView *webView = [[WKWebView alloc] initWithFrame:CGRectZero];
[webView evaluateJavaScript:... completionHandler:^{
   ...
   gBrowserUserAgent = (NSString *)result;
   ...
   dispatch_async(^{
      webView = nil;
   });
});
```

I suppose such code would have the same semantics as my variant.
Should I made such change?

Reply:
Your semantic is different: webView can be destroyed BEFORE the completion handler is called.
The idea with nil inside the completion handler+async is guaranteed to work properly.

Reply:
Why? webView is used in both blocks. In `completetionHandler ` implicitly to pass to `dispatch_async`,
and in the last `dispatch_async` explicitly to copy `keepAlive`. So always reference counter for `webView` always >= 1.

You mean, the rules of objective-c allow to remove code with side effect?

Reply:
Sorry, I'm wrong, as you already keeping the reference in the completion block. I misinterpret it as it was outside of the completion block.
IMHO, `webView = nil` is more readable, but your code is also ok.

Reply:
I changed to `webView = nil`